src/core/borrower/deployer.sol

```diff
@@ -0,0 +1,191 @@
+        collector = new Collector(shelf, pile, threshold);
+        collector.rely(msg.sender);
+        collector.deny(address(this));
+    }
+}
+
+contract PrincipalFab {
+    function newPrincipal() public returns (Principal principal) {
+        principal = new Principal();
+        principal.rely(msg.sender);
+        principal.deny(address(this));
+    }
+}
+
+contract ThresholdFab {
+    function newThreshold() public returns (PushRegistry threshold) {
+        threshold = new PushRegistry();
+        threshold.rely(msg.sender);
+        threshold.deny(address(this));
+    }
+}
+
+contract BorrowerDeployer is Auth {
+    TitleFab          titlefab;
+    LightSwitchFab    lightswitchfab;
+    ShelfFab          shelffab;
+    PileFab           pilefab;
+    PrincipalFab      principalFab;
+    CollectorFab      collectorFab;
+    ThresholdFab      thresholdFab;
+
+    address     public mainDeployer;
+
+    Title       public title;
+    LightSwitch public lightswitch;
+    Shelf       public shelf;
+    LenderLike  public lender;
+    Pile        public pile;
+    Principal   public principal;
+    Collector   public collector;
+    PushRegistry public threshold;
+
+
+    address public deployUser;
+
+
+    constructor (address mainDeployer_, TitleFab titlefab_, LightSwitchFab lightswitchfab_, ShelfFab shelffab_, PileFab pilefab_, PrincipalFab principalFab_, CollectorFab collectorFab_, ThresholdFab thresholdFab_) public {
+        deployUser = msg.sender;
+        mainDeployer = mainDeployer_;
+
+        wards[deployUser] = 1;
+        wards[mainDeployer] = 1;
+
+
+
+        titlefab = titlefab_;
+        lightswitchfab = lightswitchfab_;
+        shelffab = shelffab_;
+
+        pilefab = pilefab_;
+        principalFab = principalFab_;
+        collectorFab = collectorFab_;
+        thresholdFab = thresholdFab_;
+    }
+
+    function deployThreshold() public auth {
+        threshold = thresholdFab.newThreshold();
+        threshold.rely(mainDeployer);
+
+    }
+    function deployCollector() public auth {
+        collector = collectorFab.newCollector(address(shelf), address(pile), address(threshold));
+        collector.rely(mainDeployer);
+    }
+
+    function deployPile() public auth {
+        pile = pilefab.newPile();
+        pile.rely(mainDeployer);
+    }
+
+    function deployTitle(string memory name, string memory symbol) public auth {
+        title = titlefab.newTitle(name, symbol);
+        title.rely(mainDeployer);
+    }
+
+    function deployLightSwitch() public auth {
+        lightswitch = lightswitchfab.newLightSwitch();
+        lightswitch.rely(mainDeployer);
+    }
+
+    function deployShelf(address currency_) public auth {
+        shelf = shelffab.newShelf(currency_, address(title), address(pile), address(principal));
+        shelf.rely(mainDeployer);
+    }
+
+    function deployPrincipal() public auth {
+        principal = principalFab.newPrincipal();
+        principal.rely(mainDeployer);
+    }
+
```

Comment:
I think we are also missing the CreditLine fab here

Reply:
I think we should rename the `fabs`.

PrincipalFab and CreditLineFab should both have a method called `newCeiling`

In the deployer i would use an abstract contract called `CeilingFabLike`